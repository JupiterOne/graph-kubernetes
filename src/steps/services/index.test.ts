import { fetchServices } from '.';
import { fetchNamespaces } from '../namespaces';
import { createDataCollectionTest } from '../../../test/recording';
import { integrationConfig } from '../../../test/config';
import { Entities, Relationships } from '../constants';
import { RelationshipClass } from '@jupiterone/integration-sdk-core';

describe('#fetchServices', () => {
  test('should collect data', async () => {
    await createDataCollectionTest({
      recordingName: 'fetchServices',
      recordingDirectory: __dirname,
      integrationConfig,
      stepFunctions: [fetchNamespaces, fetchServices],
      entitySchemaMatchers: [
        {
          _type: Entities.SERVICE._type,
          matcher: {
            _class: ['Service'],
            schema: {
              additionalProperties: false,
              properties: {
                _type: { const: 'kube_service' },
                _rawData: {
                  type: 'array',
                  items: { type: 'object' },
                },
                name: { type: 'string' },
                displayName: { type: 'string' },
                generation: { type: 'number' },
                deletionGracePeriodSeconds: { type: 'string' },
                resourceVersion: { type: 'string' },
                category: {
                  type: 'array',
                  items: { type: 'string' },
                },
                function: {
                  type: 'array',
                  items: { type: 'string' },
                },
                createdOn: { type: 'number' },
                allocateLoadBalancerNodePorts: { type: 'boolean' },
                clusterIP: { type: 'string' },
                clusterIPs: {
                  type: 'array',
                  items: { type: 'string' },
                },
                externalIPs: {
                  type: 'array',
                  items: { type: 'string' },
                },
                externalName: { type: 'string' },
                externalTrafficPolicy: { type: 'string' },
                healthCheckNodePort: { type: 'number' },
                ipFamilies: {
                  type: 'array',
                  items: { type: 'string' },
                },
                ipFamilyPolicy: { type: 'string' },
                loadBalancerIP: { type: 'string' },
                loadBalancerSourceRanges: {
                  type: 'array',
                  items: { type: 'string' },
                },
                publishNotReadyAddresses: { type: 'boolean' },
                sessionAffinity: { type: 'string' },
                type: { type: 'string' },
              },
            },
          },
        },
      ],
      relationshipSchemaMatchers: [
        {
          _type: Relationships.NAMESPACE_CONTAINS_SERVICE._type,
          matcher: {
            schema: {
              properties: {
                _class: { const: RelationshipClass.CONTAINS },
                _type: { const: 'kube_namespace_contains_service' },
              },
            },
          },
        },
      ],
    });
  });
});
