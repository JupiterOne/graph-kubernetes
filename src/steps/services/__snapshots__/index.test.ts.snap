// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchServices should collect data: jobState 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "868d4703-b338-4051-98e4-b65b4f0a0006",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-01-11T15:50:15.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": Object {
                "kubernetes.io/metadata.name": "default",
              },
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:labels": Object {
                        ".": Object {},
                        "f:kubernetes.io/metadata.name": Object {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-01-11T15:50:15.000Z,
                },
              ],
              "name": "default",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "39",
              "selfLink": undefined,
              "uid": "868d4703-b338-4051-98e4-b65b4f0a0006",
            },
            "spec": V1NamespaceSpec {
              "finalizers": Array [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1704988215000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "default",
      "finalizers": Array [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "default",
      "resourceVersion": "39",
      "status.phase": "Active",
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "52615a92-75c3-4c4b-8b72-51be82a4bc21",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-01-11T15:50:15.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": Object {
                "kubernetes.io/metadata.name": "kube-node-lease",
              },
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:labels": Object {
                        ".": Object {},
                        "f:kubernetes.io/metadata.name": Object {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-01-11T15:50:15.000Z,
                },
              ],
              "name": "kube-node-lease",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "35",
              "selfLink": undefined,
              "uid": "52615a92-75c3-4c4b-8b72-51be82a4bc21",
            },
            "spec": V1NamespaceSpec {
              "finalizers": Array [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1704988215000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-node-lease",
      "finalizers": Array [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "kube-node-lease",
      "resourceVersion": "35",
      "status.phase": "Active",
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "7a2ee4f7-12fd-494b-93f0-c639b1e63a72",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-01-11T15:50:15.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": Object {
                "kubernetes.io/metadata.name": "kube-public",
              },
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:labels": Object {
                        ".": Object {},
                        "f:kubernetes.io/metadata.name": Object {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-01-11T15:50:15.000Z,
                },
              ],
              "name": "kube-public",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "20",
              "selfLink": undefined,
              "uid": "7a2ee4f7-12fd-494b-93f0-c639b1e63a72",
            },
            "spec": V1NamespaceSpec {
              "finalizers": Array [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1704988215000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-public",
      "finalizers": Array [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "kube-public",
      "resourceVersion": "20",
      "status.phase": "Active",
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "ce86116b-863b-4e74-979e-3c031a9aeee1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-01-11T15:50:15.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": Object {
                "kubernetes.io/metadata.name": "kube-system",
              },
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:labels": Object {
                        ".": Object {},
                        "f:kubernetes.io/metadata.name": Object {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-01-11T15:50:15.000Z,
                },
              ],
              "name": "kube-system",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "12",
              "selfLink": undefined,
              "uid": "ce86116b-863b-4e74-979e-3c031a9aeee1",
            },
            "spec": V1NamespaceSpec {
              "finalizers": Array [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1704988215000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-system",
      "finalizers": Array [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "kube-system",
      "resourceVersion": "12",
      "status.phase": "Active",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "7d630585-9aaa-4b35-9f38-07d7c6e5ac48",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1Service {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-01-11T15:50:16.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": Object {
                "component": "apiserver",
                "provider": "kubernetes",
              },
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:labels": Object {
                        ".": Object {},
                        "f:component": Object {},
                        "f:provider": Object {},
                      },
                    },
                    "f:spec": Object {
                      "f:clusterIP": Object {},
                      "f:internalTrafficPolicy": Object {},
                      "f:ipFamilyPolicy": Object {},
                      "f:ports": Object {
                        ".": Object {},
                        "k:{\\"port\\":443,\\"protocol\\":\\"TCP\\"}": Object {
                          ".": Object {},
                          "f:name": Object {},
                          "f:port": Object {},
                          "f:protocol": Object {},
                          "f:targetPort": Object {},
                        },
                      },
                      "f:sessionAffinity": Object {},
                      "f:type": Object {},
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-01-11T15:50:16.000Z,
                },
              ],
              "name": "kubernetes",
              "namespace": "default",
              "ownerReferences": undefined,
              "resourceVersion": "231",
              "selfLink": undefined,
              "uid": "7d630585-9aaa-4b35-9f38-07d7c6e5ac48",
            },
            "spec": V1ServiceSpec {
              "allocateLoadBalancerNodePorts": undefined,
              "clusterIP": "10.96.0.1",
              "clusterIPs": Array [
                "10.96.0.1",
              ],
              "externalIPs": undefined,
              "externalName": undefined,
              "externalTrafficPolicy": undefined,
              "healthCheckNodePort": undefined,
              "internalTrafficPolicy": "Cluster",
              "ipFamilies": Array [
                "IPv4",
              ],
              "ipFamilyPolicy": "SingleStack",
              "loadBalancerClass": undefined,
              "loadBalancerIP": undefined,
              "loadBalancerSourceRanges": undefined,
              "ports": Array [
                V1ServicePort {
                  "appProtocol": undefined,
                  "name": "https",
                  "nodePort": undefined,
                  "port": 443,
                  "protocol": "TCP",
                  "targetPort": 6443,
                },
              ],
              "publishNotReadyAddresses": undefined,
              "selector": undefined,
              "sessionAffinity": "None",
              "sessionAffinityConfig": undefined,
              "type": "ClusterIP",
            },
            "status": V1ServiceStatus {
              "conditions": undefined,
              "loadBalancer": V1LoadBalancerStatus {
                "ingress": undefined,
              },
            },
          },
        },
      ],
      "_type": "kube_service",
      "active": undefined,
      "allocateLoadBalancerNodePorts": undefined,
      "category": Array [
        "infrastructure",
      ],
      "clusterIP": "10.96.0.1",
      "clusterIPs": Array [
        "10.96.0.1",
      ],
      "createdOn": 1704988216000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kubernetes",
      "externalIPs": undefined,
      "externalName": undefined,
      "externalTrafficPolicy": undefined,
      "function": Array [
        "compute",
      ],
      "generation": undefined,
      "healthCheckNodePort": undefined,
      "ipFamilies": Array [
        "IPv4",
      ],
      "ipFamilyPolicy": "SingleStack",
      "loadBalancerIP": undefined,
      "loadBalancerSourceRanges": undefined,
      "name": "kubernetes",
      "publishNotReadyAddresses": undefined,
      "resourceVersion": "231",
      "sessionAffinity": "None",
      "type": "ClusterIP",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "de61583d-6a16-4364-a774-75225d946680",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1Service {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": Object {
                "meta.helm.sh/release-name": "np-test",
                "meta.helm.sh/release-namespace": "default",
              },
              "creationTimestamp": 2024-01-11T21:52:37.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": Object {
                "app.kubernetes.io/component": "primary",
                "app.kubernetes.io/instance": "np-test",
                "app.kubernetes.io/managed-by": "Helm",
                "app.kubernetes.io/name": "mariadb",
                "app.kubernetes.io/version": "11.2.2",
                "helm.sh/chart": "mariadb-15.0.1",
              },
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:annotations": Object {
                        ".": Object {},
                        "f:meta.helm.sh/release-name": Object {},
                        "f:meta.helm.sh/release-namespace": Object {},
                      },
                      "f:labels": Object {
                        ".": Object {},
                        "f:app.kubernetes.io/component": Object {},
                        "f:app.kubernetes.io/instance": Object {},
                        "f:app.kubernetes.io/managed-by": Object {},
                        "f:app.kubernetes.io/name": Object {},
                        "f:app.kubernetes.io/version": Object {},
                        "f:helm.sh/chart": Object {},
                      },
                    },
                    "f:spec": Object {
                      "f:internalTrafficPolicy": Object {},
                      "f:ports": Object {
                        ".": Object {},
                        "k:{\\"port\\":3306,\\"protocol\\":\\"TCP\\"}": Object {
                          ".": Object {},
                          "f:name": Object {},
                          "f:port": Object {},
                          "f:protocol": Object {},
                          "f:targetPort": Object {},
                        },
                      },
                      "f:selector": Object {},
                      "f:sessionAffinity": Object {},
                      "f:type": Object {},
                    },
                  },
                  "manager": "helm",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-01-11T21:52:37.000Z,
                },
              ],
              "name": "np-test-mariadb",
              "namespace": "default",
              "ownerReferences": undefined,
              "resourceVersion": "27438",
              "selfLink": undefined,
              "uid": "de61583d-6a16-4364-a774-75225d946680",
            },
            "spec": V1ServiceSpec {
              "allocateLoadBalancerNodePorts": undefined,
              "clusterIP": "10.103.238.222",
              "clusterIPs": Array [
                "10.103.238.222",
              ],
              "externalIPs": undefined,
              "externalName": undefined,
              "externalTrafficPolicy": undefined,
              "healthCheckNodePort": undefined,
              "internalTrafficPolicy": "Cluster",
              "ipFamilies": Array [
                "IPv4",
              ],
              "ipFamilyPolicy": "SingleStack",
              "loadBalancerClass": undefined,
              "loadBalancerIP": undefined,
              "loadBalancerSourceRanges": undefined,
              "ports": Array [
                V1ServicePort {
                  "appProtocol": undefined,
                  "name": "mysql",
                  "nodePort": undefined,
                  "port": 3306,
                  "protocol": "TCP",
                  "targetPort": "mysql",
                },
              ],
              "publishNotReadyAddresses": undefined,
              "selector": Object {
                "app.kubernetes.io/component": "primary",
                "app.kubernetes.io/instance": "np-test",
                "app.kubernetes.io/name": "mariadb",
              },
              "sessionAffinity": "None",
              "sessionAffinityConfig": undefined,
              "type": "ClusterIP",
            },
            "status": V1ServiceStatus {
              "conditions": undefined,
              "loadBalancer": V1LoadBalancerStatus {
                "ingress": undefined,
              },
            },
          },
        },
      ],
      "_type": "kube_service",
      "active": undefined,
      "allocateLoadBalancerNodePorts": undefined,
      "category": Array [
        "infrastructure",
      ],
      "clusterIP": "10.103.238.222",
      "clusterIPs": Array [
        "10.103.238.222",
      ],
      "createdOn": 1705009957000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "np-test-mariadb",
      "externalIPs": undefined,
      "externalName": undefined,
      "externalTrafficPolicy": undefined,
      "function": Array [
        "compute",
      ],
      "generation": undefined,
      "healthCheckNodePort": undefined,
      "ipFamilies": Array [
        "IPv4",
      ],
      "ipFamilyPolicy": "SingleStack",
      "loadBalancerIP": undefined,
      "loadBalancerSourceRanges": undefined,
      "name": "np-test-mariadb",
      "publishNotReadyAddresses": undefined,
      "resourceVersion": "27438",
      "sessionAffinity": "None",
      "type": "ClusterIP",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "733f1dbb-ad52-4ff2-bb2d-217f7e1ea93f",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1Service {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": Object {
                "meta.helm.sh/release-name": "np-test",
                "meta.helm.sh/release-namespace": "default",
              },
              "creationTimestamp": 2024-01-11T21:52:37.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": Object {
                "app.kubernetes.io/instance": "np-test",
                "app.kubernetes.io/managed-by": "Helm",
                "app.kubernetes.io/name": "wordpress",
                "app.kubernetes.io/version": "6.4.2",
                "helm.sh/chart": "wordpress-19.0.5",
              },
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:annotations": Object {
                        ".": Object {},
                        "f:meta.helm.sh/release-name": Object {},
                        "f:meta.helm.sh/release-namespace": Object {},
                      },
                      "f:labels": Object {
                        ".": Object {},
                        "f:app.kubernetes.io/instance": Object {},
                        "f:app.kubernetes.io/managed-by": Object {},
                        "f:app.kubernetes.io/name": Object {},
                        "f:app.kubernetes.io/version": Object {},
                        "f:helm.sh/chart": Object {},
                      },
                    },
                    "f:spec": Object {
                      "f:allocateLoadBalancerNodePorts": Object {},
                      "f:externalTrafficPolicy": Object {},
                      "f:internalTrafficPolicy": Object {},
                      "f:ports": Object {
                        ".": Object {},
                        "k:{\\"port\\":443,\\"protocol\\":\\"TCP\\"}": Object {
                          ".": Object {},
                          "f:name": Object {},
                          "f:port": Object {},
                          "f:protocol": Object {},
                          "f:targetPort": Object {},
                        },
                        "k:{\\"port\\":80,\\"protocol\\":\\"TCP\\"}": Object {
                          ".": Object {},
                          "f:name": Object {},
                          "f:port": Object {},
                          "f:protocol": Object {},
                          "f:targetPort": Object {},
                        },
                      },
                      "f:selector": Object {},
                      "f:sessionAffinity": Object {},
                      "f:type": Object {},
                    },
                  },
                  "manager": "helm",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-01-11T21:52:37.000Z,
                },
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:status": Object {
                      "f:loadBalancer": Object {
                        "f:ingress": Object {},
                      },
                    },
                  },
                  "manager": "kube-vpnkit-forwarder",
                  "operation": "Update",
                  "subresource": "status",
                  "time": 2024-01-11T21:52:37.000Z,
                },
              ],
              "name": "np-test-wordpress",
              "namespace": "default",
              "ownerReferences": undefined,
              "resourceVersion": "27471",
              "selfLink": undefined,
              "uid": "733f1dbb-ad52-4ff2-bb2d-217f7e1ea93f",
            },
            "spec": V1ServiceSpec {
              "allocateLoadBalancerNodePorts": true,
              "clusterIP": "10.108.56.15",
              "clusterIPs": Array [
                "10.108.56.15",
              ],
              "externalIPs": undefined,
              "externalName": undefined,
              "externalTrafficPolicy": "Cluster",
              "healthCheckNodePort": undefined,
              "internalTrafficPolicy": "Cluster",
              "ipFamilies": Array [
                "IPv4",
              ],
              "ipFamilyPolicy": "SingleStack",
              "loadBalancerClass": undefined,
              "loadBalancerIP": undefined,
              "loadBalancerSourceRanges": undefined,
              "ports": Array [
                V1ServicePort {
                  "appProtocol": undefined,
                  "name": "http",
                  "nodePort": 30766,
                  "port": 80,
                  "protocol": "TCP",
                  "targetPort": "http",
                },
                V1ServicePort {
                  "appProtocol": undefined,
                  "name": "https",
                  "nodePort": 31472,
                  "port": 443,
                  "protocol": "TCP",
                  "targetPort": "https",
                },
              ],
              "publishNotReadyAddresses": undefined,
              "selector": Object {
                "app.kubernetes.io/instance": "np-test",
                "app.kubernetes.io/name": "wordpress",
              },
              "sessionAffinity": "None",
              "sessionAffinityConfig": undefined,
              "type": "LoadBalancer",
            },
            "status": V1ServiceStatus {
              "conditions": undefined,
              "loadBalancer": V1LoadBalancerStatus {
                "ingress": Array [
                  V1LoadBalancerIngress {
                    "hostname": "localhost",
                    "ip": undefined,
                    "ports": undefined,
                  },
                ],
              },
            },
          },
        },
      ],
      "_type": "kube_service",
      "active": undefined,
      "allocateLoadBalancerNodePorts": true,
      "category": Array [
        "infrastructure",
      ],
      "clusterIP": "10.108.56.15",
      "clusterIPs": Array [
        "10.108.56.15",
      ],
      "createdOn": 1705009957000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "np-test-wordpress",
      "externalIPs": undefined,
      "externalName": undefined,
      "externalTrafficPolicy": "Cluster",
      "function": Array [
        "compute",
      ],
      "generation": undefined,
      "healthCheckNodePort": undefined,
      "ipFamilies": Array [
        "IPv4",
      ],
      "ipFamilyPolicy": "SingleStack",
      "loadBalancerIP": undefined,
      "loadBalancerSourceRanges": undefined,
      "name": "np-test-wordpress",
      "publishNotReadyAddresses": undefined,
      "resourceVersion": "27471",
      "sessionAffinity": "None",
      "type": "LoadBalancer",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "f6f4c86d-d361-41b0-931d-58df2f0324cc",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1Service {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": Object {
                "prometheus.io/port": "9153",
                "prometheus.io/scrape": "true",
              },
              "creationTimestamp": 2024-01-11T15:50:18.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": Object {
                "k8s-app": "kube-dns",
                "kubernetes.io/cluster-service": "true",
                "kubernetes.io/name": "CoreDNS",
              },
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:annotations": Object {
                        ".": Object {},
                        "f:prometheus.io/port": Object {},
                        "f:prometheus.io/scrape": Object {},
                      },
                      "f:labels": Object {
                        ".": Object {},
                        "f:k8s-app": Object {},
                        "f:kubernetes.io/cluster-service": Object {},
                        "f:kubernetes.io/name": Object {},
                      },
                    },
                    "f:spec": Object {
                      "f:clusterIP": Object {},
                      "f:internalTrafficPolicy": Object {},
                      "f:ports": Object {
                        ".": Object {},
                        "k:{\\"port\\":53,\\"protocol\\":\\"TCP\\"}": Object {
                          ".": Object {},
                          "f:name": Object {},
                          "f:port": Object {},
                          "f:protocol": Object {},
                          "f:targetPort": Object {},
                        },
                        "k:{\\"port\\":53,\\"protocol\\":\\"UDP\\"}": Object {
                          ".": Object {},
                          "f:name": Object {},
                          "f:port": Object {},
                          "f:protocol": Object {},
                          "f:targetPort": Object {},
                        },
                        "k:{\\"port\\":9153,\\"protocol\\":\\"TCP\\"}": Object {
                          ".": Object {},
                          "f:name": Object {},
                          "f:port": Object {},
                          "f:protocol": Object {},
                          "f:targetPort": Object {},
                        },
                      },
                      "f:selector": Object {},
                      "f:sessionAffinity": Object {},
                      "f:type": Object {},
                    },
                  },
                  "manager": "kubeadm",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-01-11T15:50:18.000Z,
                },
              ],
              "name": "kube-dns",
              "namespace": "kube-system",
              "ownerReferences": undefined,
              "resourceVersion": "275",
              "selfLink": undefined,
              "uid": "f6f4c86d-d361-41b0-931d-58df2f0324cc",
            },
            "spec": V1ServiceSpec {
              "allocateLoadBalancerNodePorts": undefined,
              "clusterIP": "10.96.0.10",
              "clusterIPs": Array [
                "10.96.0.10",
              ],
              "externalIPs": undefined,
              "externalName": undefined,
              "externalTrafficPolicy": undefined,
              "healthCheckNodePort": undefined,
              "internalTrafficPolicy": "Cluster",
              "ipFamilies": Array [
                "IPv4",
              ],
              "ipFamilyPolicy": "SingleStack",
              "loadBalancerClass": undefined,
              "loadBalancerIP": undefined,
              "loadBalancerSourceRanges": undefined,
              "ports": Array [
                V1ServicePort {
                  "appProtocol": undefined,
                  "name": "dns",
                  "nodePort": undefined,
                  "port": 53,
                  "protocol": "UDP",
                  "targetPort": 53,
                },
                V1ServicePort {
                  "appProtocol": undefined,
                  "name": "dns-tcp",
                  "nodePort": undefined,
                  "port": 53,
                  "protocol": "TCP",
                  "targetPort": 53,
                },
                V1ServicePort {
                  "appProtocol": undefined,
                  "name": "metrics",
                  "nodePort": undefined,
                  "port": 9153,
                  "protocol": "TCP",
                  "targetPort": 9153,
                },
              ],
              "publishNotReadyAddresses": undefined,
              "selector": Object {
                "k8s-app": "kube-dns",
              },
              "sessionAffinity": "None",
              "sessionAffinityConfig": undefined,
              "type": "ClusterIP",
            },
            "status": V1ServiceStatus {
              "conditions": undefined,
              "loadBalancer": V1LoadBalancerStatus {
                "ingress": undefined,
              },
            },
          },
        },
      ],
      "_type": "kube_service",
      "active": undefined,
      "allocateLoadBalancerNodePorts": undefined,
      "category": Array [
        "infrastructure",
      ],
      "clusterIP": "10.96.0.10",
      "clusterIPs": Array [
        "10.96.0.10",
      ],
      "createdOn": 1704988218000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-dns",
      "externalIPs": undefined,
      "externalName": undefined,
      "externalTrafficPolicy": undefined,
      "function": Array [
        "compute",
      ],
      "generation": undefined,
      "healthCheckNodePort": undefined,
      "ipFamilies": Array [
        "IPv4",
      ],
      "ipFamilyPolicy": "SingleStack",
      "loadBalancerIP": undefined,
      "loadBalancerSourceRanges": undefined,
      "name": "kube-dns",
      "publishNotReadyAddresses": undefined,
      "resourceVersion": "275",
      "sessionAffinity": "None",
      "type": "ClusterIP",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "CONTAINS",
      "_fromEntityKey": "868d4703-b338-4051-98e4-b65b4f0a0006",
      "_key": "868d4703-b338-4051-98e4-b65b4f0a0006|contains|7d630585-9aaa-4b35-9f38-07d7c6e5ac48",
      "_toEntityKey": "7d630585-9aaa-4b35-9f38-07d7c6e5ac48",
      "_type": "kube_namespace_contains_service",
      "displayName": "CONTAINS",
    },
    Object {
      "_class": "CONTAINS",
      "_fromEntityKey": "868d4703-b338-4051-98e4-b65b4f0a0006",
      "_key": "868d4703-b338-4051-98e4-b65b4f0a0006|contains|de61583d-6a16-4364-a774-75225d946680",
      "_toEntityKey": "de61583d-6a16-4364-a774-75225d946680",
      "_type": "kube_namespace_contains_service",
      "displayName": "CONTAINS",
    },
    Object {
      "_class": "CONTAINS",
      "_fromEntityKey": "868d4703-b338-4051-98e4-b65b4f0a0006",
      "_key": "868d4703-b338-4051-98e4-b65b4f0a0006|contains|733f1dbb-ad52-4ff2-bb2d-217f7e1ea93f",
      "_toEntityKey": "733f1dbb-ad52-4ff2-bb2d-217f7e1ea93f",
      "_type": "kube_namespace_contains_service",
      "displayName": "CONTAINS",
    },
    Object {
      "_class": "CONTAINS",
      "_fromEntityKey": "ce86116b-863b-4e74-979e-3c031a9aeee1",
      "_key": "ce86116b-863b-4e74-979e-3c031a9aeee1|contains|f6f4c86d-d361-41b0-931d-58df2f0324cc",
      "_toEntityKey": "f6f4c86d-d361-41b0-931d-58df2f0324cc",
      "_type": "kube_namespace_contains_service",
      "displayName": "CONTAINS",
    },
  ],
  "encounteredTypes": Array [
    "kube_namespace",
    "kube_service",
    "kube_namespace_contains_service",
  ],
  "numCollectedEntities": 8,
  "numCollectedRelationships": 4,
}
`;
